src/ocnicepost.fd/ocnicepost.F90

```diff
@@ -0,0 +1,577 @@
+  read (nml=ocnicepost_nml, iostat=rc, unit=iounit)
+  if (rc /= 0) then
+     write (6, '(a)') 'Error: invalid Namelist format.'
+  end if
+  close (iounit)
+
+  ! initialize the source file type and variables
+  if (trim(ftype) == 'ocean') then
+     do_ocnpost = .true.
+     call ocnvars_typedefine
+  else
+     do_ocnpost = .false.
+     call icevars_typedefine
+  end if
+  input_file = trim(ftype)//'.nc'
+
+  open(newunit=logunit, file=trim(ftype)//'.post.log',form='formatted')
+  if (debug) write(logunit, '(a)')'input file: '//trim(input_file)
+
+  ! --------------------------------------------------------
+  ! read the source file and obtain the units and long name,
+  ! rotation angles, vertical grid and time axis
+  ! --------------------------------------------------------
+
+  nvalid = 0
+  rc = nf90_open(trim(input_file), nf90_nowrite, ncid)
+  do i = 1,maxvars
+     if (len_trim(outvars(i)%input_var_name) > 0 ) then
+        rc = nf90_inq_varid(ncid, trim(outvars(i)%input_var_name), varid)
+        rc = nf90_get_att(ncid, varid,  'long_name', outvars(i)%long_name)
+        rc = nf90_get_att(ncid, varid,      'units', outvars(i)%units)
+        rc = nf90_get_att(ncid, varid, '_FillValue', outvars(i)%var_fillvalue)
+        nvalid = nvalid+1
+        if (trim(outvars(i)%input_var_name) ==   'temp')vfill = outvars(i)%var_fillvalue
+        if (trim(outvars(i)%input_var_name) == 'aice_h')vfill = outvars(i)%var_fillvalue
+     end if
+  end do
+
+  ! timestamp
+  rc = nf90_inq_varid(ncid, 'time', varid)
+  rc = nf90_get_var(ncid, varid, timestamp)
+  rc = nf90_get_att(ncid, varid,    'units', timeunit)
+  rc = nf90_get_att(ncid, varid, 'calendar', timecal)
+  if (do_ocnpost) then
+     ! cell centers
+     rc = nf90_inq_varid(ncid, 'z_l', varid)
+     rc = nf90_get_var(ncid, varid, z_l)
+     ! cell edges
+     rc = nf90_inq_varid(ncid, 'z_i', varid)
+     rc = nf90_get_var(ncid, varid, z_i)
+     rc = nf90_close(ncid)
+     ! rotation angles
+     call getfield(trim(input_file), 'cos_rot', dims=(/nxt,nyt/), field=cosrot)
+     call getfield(trim(input_file), 'sin_rot', dims=(/nxt,nyt/), field=sinrot)
+  else
+     call getfield(trim(input_file),  'ANGLET', dims=(/nxt,nyt/), field=anglet)
+     cosrot =  cos(anglet)
+     sinrot = -sin(anglet)
+  end if
+
+  ! --------------------------------------------------------
+  ! mask is a 2d (ice) or 3d (ocn) array which contains 1's
+  ! on land and 0's at valid points.
+  ! when remapped, any mask value > 0 identifies land values that
+  ! have crept into the field. remapped model fields are then
+  ! masked with this interpolation mask
+  ! --------------------------------------------------------
+
+  if (do_ocnpost) then
+     rc = nf90_open(trim(input_file), nf90_nowrite, ncid)
+     ! 3D temp to use as mask, obtain directly from file to preserve vfill
+     rc = nf90_inq_varid(ncid, 'temp', varid)
+     rc = nf90_get_var(ncid, varid, tmp3d)
```

Comment:
why not use `getfield`?

Reply:
The getfield is not used so that I can retain the fill value, which I use to define the 3-D interpolation mask.